data/tokenizer.py

```diff
@@ -237,3 +237,103 @@ def untokenize_tgt(self, tokens: list[str]) -> list[int]:
         starts = [int(token[6:]) for token in tokens if token not in self.specials]
 
         return starts
+
+
+class MaskedMidiEncoder(MultiTokEncoder):
+    def __init__(self, quantization_cfg: DictConfig, keys: list[str] = None, masking_probability: float = 0.1):
+        super().__init__()
+        self.quantization_cfg = quantization_cfg
+        self.masking_probability = masking_probability
+        if keys is None:
+            self.keys = ["pitch", "start_bin", "duration_bin", "velocity_bin"]
+        else:
+            self.keys = keys
+
+        self.specials = ["<PAD>", "<MASK>"]
+
+        self.vocab = list(self.specials)
+
+        # add midi tokens to vocab
+        self._build_vocab()
+        self.token_to_id = {token: it for it, token in enumerate(self.vocab)}
+
+    def __rich_repr__(self):
+        yield "MaskedMidiEncoder"
+        yield "vocab_size", self.vocab_size
+
+    @property
+    def vocab_size(self) -> int:
+        return len(self.vocab)
+
+    def _build_vocab(self):
+        time_tokens_product = itertools.product(
+            # weird because we want to use dstart if there is a dstart_bin key, and start if there is a start_bin
+            range(self.quantization_cfg[self.keys[1][:-4]]),
+            range(self.quantization_cfg.duration),
```

Comment:
Please make it not weird and easy to understand :)

If this is checking if the tokenizer should use `start` or `dstart`, it should be possible to be explicit about it

Reply:
it generalizes it to both scenarios, it is possible to use dstart or start now, I can make it with ifs instead if you want :)

Reply:
I think we should be able to be more explicit about this. I also don't like the way `keys` are used in this class - we should find a better name (what type of keys are those), and make this a required argument, so get rid of the `if keys is None` part in the `__init__`.

Then getting the number of bins related to time could look like this:

```py
self.quantization_cfg[self.time_quantization_method]
```

And quantization method would ba aproperty inferred from the keys argument.

Reply:
done, you can check out if the way it is done now looks good enough for you